### PR TITLE
Fix relative urls & source maps

### LINKED ISSUE #284: Source maps for files in sub-directories don't work
When playing with Sprockets 4.0.0.beta2 on one of our apps, I noticed a bug which leads to sourcemaps not working.

When I load an asset file from a subdirectory (`javascript_include_tag "sub/directory"`), the file contains a sourcemaps entry with a relative path duplicating the `sub` path:

```javascript
/# sourceMappingURL=sub/directory.js-SHA.map
```

This makes my chrome devtools request `/assets/sub/sub/directory.js-SHA.map` which of course fails.

To make sure it's not our (rather huge) application itself, I created a small test rails application to demonstrate the issue at [siebertm/sprockets-sourcemaps-bug](https://github.com/siebertm/sprockets-sourcemaps-bug)

This is a standard rails 5.0.0.beta3 `rails new` generated application. The offending files are [app/views/layouts/application.html.erb:10](https://github.com/siebertm/sprockets-sourcemaps-bug/blob/master/app/views/layouts/application.html.erb#L10) and [app/assets/javascripts/sub/directory.js](https://github.com/siebertm/sprockets-sourcemaps-bug/blob/master/app/assets/javascripts/sub/directory.js)

The default `application.js` sourcemap works like a charm!

Is this indeed a bug or am I misunderstanding anything here?

### PR DESCRIPTION
**Pull request by TylerHorth**
This PR fixes two issues related to source maps and relative URLs. 

1. As reported in #284, the `SourceMapCommentProcessor` uses the logical path as the source mapping url which is relative to the load path. However, from the [source map revision 3 proposal](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit), the url should be relative to the `.debug` file itself. This meant that source maps didn't work when loading assets not located in the assets root directory. 
2. A similar issue to the first popped up within the `.map` files. Sources should be linked relative to the `.map` file itself, however again they were being linked relative to the load path. 

I don't believe I have the best solution for this problem, specifically the hack in `source_map_utils.rb`, and it's likely that there are edge cases where it breaks down, but it's necessary to get source maps functioning. I'd love to hear suggestions on ways to improve this solution. 

fixes #284 

@rafaelfranca @bouk

### CODE REVIEW

**Comment by TylerHorth** [comment]
Looks like I need to fix the tests.

----
```
  1) Failure:
TestSourceMaps#test_"compile babel source map" [C:/projects/sprockets/test/test_source_maps.rb:137]:
--- expected
+++ actual
@@ -1 +1 @@
-{"version"=>3, "file"=>"babel/main.js", "mappings"=>";;;;;;;;;;AACA,IAAI,IAAI,GAAG,KAAK,CAAC,GAAG,CAAC,UAAA,CAAC;SAAI,CAAC,GAAG,CAAC;CAAA,CAAC,CAAC;AACjC,IAAI,IAAI,GAAG,KAAK,CAAC,GAAG,CAAC,UAAC,CAAC,EAAE,CAAC;SAAK,CAAC,GAAG,CAAC;CAAA,CAAC,CAAC;;IAEhC,WAAW;YAAX,WAAW;;AACJ,WADP,WAAW,CACH,QAAQ,EAAE,SAAS,EAAE;0BAD7B,WAAW;;AAEb,+BAFE,WAAW,6CAEP,QAAQ,EAAE,SAAS,EAAE;GAE5B;;eAJG,WAAW;;WAKT,gBAAC,MAAM,EAAE;AACb,iCANE,WAAW,wCAME;KAChB;;;WACmB,yBAAG;AACrB,aAAO,IAAI,KAAK,CAAC,OAAO,EAAE,CAAC;KAC5B;;;SAVG,WAAW;GAAS,KAAK,CAAC,IAAI;;AAapC,IAAI,SAAS,uBACV,MAAM,CAAC,QAAQ,0BAAG;MACb,GAAG,EAAM,GAAG,EAEV,IAAI;;;;AAFN,WAAG,GAAG,CAAC,EAAE,GAAG,GAAG,CAAC;;;AAEd,YAAI,GAAG,GAAG;;AACd,WAAG,GAAG,GAAG,CAAC;AACV,WAAG,IAAI,IAAI,CAAC;;eACN,GAAG;;;;;;;;;;;CAEZ,EACF,CAAA", "sources"=>["babel/main.source-1acb9cf16a3e1ce0fe0a38491472a14a6a97281ceace4b67ec16a904be5fa1b9.es6"], "names"=>[]}
+{"version"=>3, "file"=>"babel/main.js", "mappings"=>";;;;;;;;;;AACA,IAAI,IAAI,GAAG,KAAK,CAAC,GAAG,CAAC,UAAA,CAAC;SAAI,CAAC,GAAG,CAAC;CAAA,CAAC,CAAC;AACjC,IAAI,IAAI,GAAG,KAAK,CAAC,GAAG,CAAC,UAAC,CAAC,EAAE,CAAC;SAAK,CAAC,GAAG,CAAC;CAAA,CAAC,CAAC;;IAEhC,WAAW;YAAX,WAAW;;AACJ,WADP,WAAW,CACH,QAAQ,EAAE,SAAS,EAAE;0BAD7B,WAAW;;AAEb,+BAFE,WAAW,6CAEP,QAAQ,EAAE,SAAS,EAAE;GAE5B;;eAJG,WAAW;;WAKT,gBAAC,MAAM,EAAE;AACb,iCANE,WAAW,wCAME;KAChB;;;WACmB,yBAAG;AACrB,aAAO,IAAI,KAAK,CAAC,OAAO,EAAE,CAAC;KAC5B;;;SAVG,WAAW;GAAS,KAAK,CAAC,IAAI;;AAapC,IAAI,SAAS,uBACV,MAAM,CAAC,QAAQ,0BAAG;MACb,GAAG,EAAM,GAAG,EAEV,IAAI;;;;AAFN,WAAG,GAAG,CAAC,EAAE,GAAG,GAAG,CAAC;;;AAEd,YAAI,GAAG,GAAG;;AACd,WAAG,GAAG,GAAG,CAAC;AACV,WAAG,IAAI,IAAI,CAAC;;eACN,GAAG;;;;;;;;;;;CAEZ,EACF,CAAA", "sources"=>["main.source-1acb9cf16a3e1ce0fe0a38491472a14a6a97281ceace4b67ec16a904be5fa1b9.es6"], "names"=>[]}
```
This demonstrates the second issue. The `sources` should actually be relative to the `file`. From the [proposition](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit#heading=h.75yo6yoyk7x5):
> If the sources are not absolute URLs after prepending of the “sourceRoot”, the sources are resolved relative to the SourceMap (like resolving script src in a html document).
In this case the source map file is `babel/main.js.map`, so a source of `babel/main.source.es6` will resolve to `babel/babel/main.source.es6`.

**Comment by schneems** [comment]
> I don't believe I have the best solution for this problem, specifically the hack in source_map_utils.rb, and it's likely that there are edge cases where it breaks down

I took a look at your solution and don't see a better way. My approach with source maps has been to fix bugs, add tests for them, and cut a beta release. We then wait to see if other people report bugs, and fix those without test regressions. Hopefully we'll catch the majority of edge cases before a RC.

I think i'm fine with this as is. Can you modify that failing test so that it passes? Since that is testing the case you're trying to handle, we don't need to add additional test cases.

**Comment by schneems** [comment]
Getting an error in the 2.3 tests

```
TestSourceMapUtils#test_map2:
NameError: uninitialized constant Sprockets::SourceMapUtils::PathUtils
    /home/travis/build/rails/sprockets/lib/sprockets/source_map_utils.rb:140:in `block in encode_json_source_map'
    /home/travis/build/rails/sprockets/lib/sprockets/source_map_utils.rb:139:in `each'
    /home/travis/build/rails/sprockets/lib/sprockets/source_map_utils.rb:139:in `encode_json_source_map'
    test/test_source_map_utils.rb:66:in `test_map2'
```

**Comment by TylerHorth** [comment]
Strange that it's happening on 2.3.0 but not 2.3.1. Looking into it.

**Comment by schneems** [comment]
Agreed, i restarted the test to make sure there wasn't a hiccup and it is still failing.

**Comment by schneems** [comment]
Looks like that did it. Not sure why it failed in 2.3.0 but not 2.3.1. Pretty weird. Thanks for your work 🚀
